File: unisync_android/app/src/main/java/com/anhquan/unisync/database/dao/PairedDeviceDao.kt
package com.anhquan.unisync.database.dao

import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import com.anhquan.unisync.database.entity.PairedDeviceEntity
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.Single

@Dao
interface PairedDeviceDao {
    @Query("SELECT * FROM paired_devices")
    fun getAll(): Single<List<PairedDeviceEntity>>

    @Query("SELECT * FROM paired_devices WHERE id = :id")
    fun get(id: String): Single<PairedDeviceEntity>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun add(value: PairedDeviceEntity): Completable

    @Delete
    fun remove(value: PairedDeviceEntity): Completable

    @Query("DELETE FROM paired_devices WHERE id = :id")
    fun remove(id: String): Completable

    @Query("SELECT COUNT(id) FROM paired_devices WHERE id = :id")
    fun exist(id: String): Single<Int>

    @Query("SELECT * FROM paired_devices ORDER BY lastAccessed DESC LIMIT 1")
    fun getLastUsed(): Single<PairedDeviceEntity>

    @Query("UPDATE paired_devices SET lastAccessed = :lastAccessed WHERE id = :id")
    fun setLastUsed(id: String, lastAccessed: Long = System.currentTimeMillis()): Completable
}

File: lib/database/dao/paired_device.dao.dart
import 'package:floor/floor.dart';
import 'package:unisync/database/entity/paired_device.entity.dart';
import 'package:unisync/database/tables.dart';

@dao
abstract class PairedDeviceDao {
  @Query('SELECT * FROM paired_devices')
  Future<List<PairedDeviceEntity>> getAll();

  @Query('SELECT * FROM paired_devices WHERE id = :id')
  Future<PairedDeviceEntity?> get(String id);

  @Insert(onConflict: OnConflictStrategy.replace)
  Future<void> add(PairedDeviceEntity device);

  @Query('UPDATE ${DBTables.PAIRED_DEVICES} SET unpaired = true WHERE id = :id')
  Future<void> markUnpaired(String id);

  @Query('DELETE FROM ${DBTables.PAIRED_DEVICES} WHERE id = :id')
  Future<void> remove(String id);

  @Query('SELECT COUNT(id) FROM ${DBTables.PAIRED_DEVICES} WHERE id = :id')
  Future<int?> exist(String id);

  @Query('SELECT * FROM ${DBTables.PAIRED_DEVICES} ORDER BY lastAccessed DESC LIMIT 1')
  Future<PairedDeviceEntity?> getLastUsed();

  @Query('UPDATE ${DBTables.PAIRED_DEVICES} SET lastAccessed = :lastAccessed WHERE id = :id')
  Future<void> setLastUsed(String id, int lastAccessed);
}
